src/pages/instances/actions/ConvertToInstanceBtn.tsx

```diff
@@ -0,0 +1,321 @@
+    const message = (
+      <>
+        Created instance <strong>{instanceName}</strong>.
+      </>
+    );
+
+    const actions = [
+      {
+        label: "Configure",
+        onClick: () =>
+          navigate(
+            `/ui/project/${project?.name}/instance/${instanceName}/configuration`,
+          ),
+      },
+    ];
+
+    toastNotify.success(message, actions);
+  };
+
+  const handleFailure = (msg: string) => {
+    toastNotify.failure("Instance creation failed.", new Error(msg));
+  };
+
+  const handleFinish = () => {
+    void queryClient.invalidateQueries({
+      predicate: (query) => {
+        return query.queryKey[0] === queryKeys.instances;
+      },
+    });
+  };
+
+  const handleConfirmUpload = (values: ConvertToInstanceFormValues) => {
+    if (!values.imageFile) {
+      return;
+    }
+
+    // start create instance operation
+    createInstance(convertToInstancePayload(values), project?.name || "")
+      .then((operation) => {
+        const operationId = operation.metadata.id;
+        const operationSecret = operation.metadata?.metadata?.fs;
+
+        // establish websocket connection based on the instance creation operation
+        let wsUrl = `wss://${location.host}/1.0/operations/${operationId}/websocket`;
+        if (operationSecret) {
+          wsUrl += `?secret=${operationSecret}`;
+        }
+
+        const ws = new WebSocket(wsUrl);
+
+        ws.onopen = () => {
+          sendFileByWebSocket(
+            ws,
+            formik.values.imageFile,
+            CHUNK_SIZE,
+            handleUploadProgress,
+            handleUploadError,
+          );
+        };
+
+        // set up event queue for the operation
+        eventQueue.set(
+          operationId,
+          () => handleSuccess(getInstanceName(operation.metadata)),
+          handleFailure,
+          handleFinish,
+        );
+      })
+      .catch((e) => {
+        handleCloseModal();
+        toastNotify.failure("Instance creation failed.", e);
+      });
+  };
+
+  const architectureOptions = getSupportedArchitectures("virtual-machine").map(
+    (arch) => ({ value: arch, label: arch }),
+  );
+
+  const formik = useFormik<ConvertToInstanceFormValues>({
+    initialValues: {
+      name: "",
+      pool: "",
+      imageFile: null,
+      formatConversion: true,
+      virtioConversion: false,
+      architecture: architectureOptions[0].value,
+    },
+    validateOnMount: true,
+    validationSchema: Yup.object().shape({
+      name: instanceNameValidation(
+        project?.name || "",
+        instanceNameAbort,
+      ).optional(),
+    }),
+    onSubmit: handleConfirmUpload,
+  });
+
+  const handleCloseModal = useCallback(() => {
+    formik.resetForm();
+    closePortal();
+  }, [formik.resetForm, closePortal]);
+
+  return (
+    <>
+      <Button onClick={openPortal} type="button">
+        <span>Convert from image</span>
+      </Button>
+      {isOpen && (
+        <Portal>
+          <Modal
```

Comment:
Maybe split the modal and the upload logic into a dedicated component?

Reply:
Sounds good, I've moved all util functions into `util/convertToInstance.tsx`, moved the form component into `ConvertToInstanceForm.tsx` and kept the upload logic in `ConvertToInstanceBtn.tsx`.